libecole/src/reward/boundintegral.cpp

```diff
@@ -196,92 +240,128 @@ auto compute_primal_dual_integral(
 	assert(handler != nullptr);
 	return *handler;
 }
 
 /** Add the integral event handler to the model. */
-void add_eventhdlr(scip::Model& model, bool wall, bool extract_primal, bool extract_dual) {
-	auto handler = std::make_unique<IntegralEventHandler>(model.get_scip_ptr(), wall, extract_primal, extract_dual);
+void add_eventhdlr(scip::Model& model, bool wall, bool extract_primal, bool extract_dual, const char* name) {
+	auto handler = std::make_unique<IntegralEventHandler>(model.get_scip_ptr(), wall, extract_primal, extract_dual, name);
 	scip::call(SCIPincludeObjEventhdlr, model.get_scip_ptr(), handler.get(), true);
 	// NOLINTNEXTLINE memory ownership is passed to SCIP
 	handler.release();
 	// NOLINTNEXTLINE memory ownership is passed to SCIP
 }
 
 /** Default function for returning +/-infinity for the bounds in computing primal-dual integral. */
-auto default_bound_function(scip::Model& model) -> std::tuple<Reward, Reward> {
+auto default_dual_bound_function(scip::Model& model) -> std::tuple<Reward, Reward> {
+	if (SCIPgetObjsense(model.get_scip_ptr()) == SCIP_OBJSENSE_MINIMIZE) {
+		return {0.0, -SCIPinfinity(model.get_scip_ptr())};
+	}
+	return {0.0, SCIPinfinity(model.get_scip_ptr())};
+}
+
+/** Default function for returning +/-infinity for the bounds in computing primal-dual integral. */
+auto default_primal_bound_function(scip::Model& model) -> std::tuple<Reward, Reward> {
+	if (SCIPgetObjsense(model.get_scip_ptr()) == SCIP_OBJSENSE_MINIMIZE) {
+		return {0.0, SCIPinfinity(model.get_scip_ptr())};
+	}
+	return {0.0, -SCIPinfinity(model.get_scip_ptr())};
+}
+
+/** Default function for returning +/-infinity for the bounds in computing primal-dual integral. */
+auto default_primal_dual_bound_function(scip::Model& model) -> std::tuple<Reward, Reward> {
+	if (SCIPgetObjsense(model.get_scip_ptr()) == SCIP_OBJSENSE_MINIMIZE) {
+		return {SCIPinfinity(model.get_scip_ptr()), -SCIPinfinity(model.get_scip_ptr())};
+	}
 	return {-SCIPinfinity(model.get_scip_ptr()), SCIPinfinity(model.get_scip_ptr())};
 }
 
 }  // namespace
 
 template <Bound bound>
-ecole::reward::BoundIntegral<bound>::BoundIntegral(bool wall_, const BoundFunction& bound_function_) :
-	bound_function{bound_function_ ? bound_function_ : default_bound_function}, wall{wall_} {}
+ecole::reward::BoundIntegral<bound>::BoundIntegral(bool wall_, const BoundFunction& bound_function_) : wall{wall_} {
+	if constexpr (bound == Bound::dual) {
+		bound_function = bound_function_ ? bound_function_ : default_dual_bound_function;
+	} else if constexpr (bound == Bound::primal) {
+		bound_function = bound_function_ ? bound_function_ : default_primal_bound_function;
+	} else if constexpr (bound == Bound::primal_dual) {
+		bound_function = bound_function_ ? bound_function_ : default_primal_dual_bound_function;
+	}
+	name = IntegralEventHandler::base_name + std::to_string(IntegralEventHandler::integral_reward_function_counter);
+	IntegralEventHandler::integral_reward_function_counter++;
```

Comment:
Should we add a lock here? (see `Scimpl::copy` for an example).

Reply:
Yea I think it makes sense to add one here.  It might not really be needed since different threads would have different instances, leading to no name conflicts, but I think it is more intuitive to make each reward function have a uniquely named event handler anyway.

Reply:
I think ++ should is an atomic operation for ints...
Maybe using std::atomic would make sure of it ?
https://en.cppreference.com/w/cpp/atomic/atomic

Reply:
I don't think that would resolve the case where two threads give `name` the same value since it could be the case that ++ may not be called in either thread right?

Reply:
Never mind, the whole action of reading and then incrementing the counter should be atomic. Looks good now.